test/common/grpc/async_client_impl_test.cc

```diff
@@ -0,0 +1,238 @@
+#include "common/grpc/async_client_impl.h"
+
+#include "test/mocks/buffer/mocks.h"
+#include "test/mocks/grpc/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/proto/helloworld.pb.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+using testing::_;
+using testing::Invoke;
+using testing::Eq;
+using testing::NiceMock;
+using testing::Return;
+using testing::ReturnRef;
+using testing::Mock;
+
+namespace Grpc {
+namespace {
+
+const std::string HELLO_REQUEST = "ABC";
+// We expect the 5 byte header to only have a length of 5 indicating the size of the protobuf. The
+// protobuf beings with 0xa, indicating this is the first field of type string. This is followed
+// by 0x03 for the number of characters and the name ABC set above.
+const char HELLO_REQUEST_DATA[] = "\x00\x00\x00\x00\x05\x0a\x03\x41\x42\x43";
+const size_t HELLO_REQUEST_SIZE = sizeof(HELLO_REQUEST_DATA) - 1;
+
+const std::string HELLO_REPLY = "CBA";
```

Comment:
I would pick HELLO_REPLY as something like 'DEFG', so HELLO_REPLY_SIZE will be different with HELLO_REQUEST_SIZE. And the content will be different as well.